components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java

```diff
@@ -439,27 +470,50 @@ public BlobRef setBlob(final InputStream inputStream, final Map<String, String>
     BlobRef oldBlobRef = asset.blobRef();
     if (oldBlobRef != null) {
       if (effectiveWritePolicy == WritePolicy.ALLOW_ONCE) {
-        throw new IllegalOperationException("Repository does not allow updating assets: " + getBucket().repositoryName());
+        throw new IllegalOperationException(
+            "Repository does not allow updating assets: " + getBucket().repositoryName());
       }
       deleteBlob(oldBlobRef, effectiveWritePolicy);
     }
 
-    // Store new blob while calculating hashes in one pass
-    final MultiHashingInputStream hashingStream = new MultiHashingInputStream(hashAlgorithms, inputStream);
-    final BlobRef newBlobRef = createBlob(hashingStream, headers);
-
-    asset.blobRef(newBlobRef);
-    asset.size(hashingStream.count());
-    asset.contentType(contentType);
+    asset.blobRef(assetBlob.getBlobRef());
+    asset.size(assetBlob.getSize());
+    asset.contentType(assetBlob.getContentType());
 
     // Set attributes map to contain computed checksum metadata
-    Map<HashAlgorithm, HashCode> hashes = hashingStream.hashes();
     NestedAttributesMap checksums = asset.attributes().child(P_CHECKSUM);
-    for (HashAlgorithm algorithm : hashAlgorithms) {
-      checksums.set(algorithm.name(), hashes.get(algorithm).toString());
+    for (HashAlgorithm algorithm : assetBlob.getHashes().keySet()) {
+      checksums.set(algorithm.name(), assetBlob.getHashes().get(algorithm).toString());
     }
 
-    return newBlobRef;
+    assetBlob.setAttached(true);
+  }
+
+  @Override
+  public BlobRef setBlob(final Asset asset,
+                         final String blobNameHint,
+                         final InputStream inputStream,
+                         final Iterable<HashAlgorithm> hashAlgorithms,
+                         @Nullable final Map<String, String> headers,
+                         @Nullable final String declaredContentType)
+  {
+    checkNotNull(asset);
+
+    // Enforce write policy ahead, as we have asset here
+    BlobRef oldBlobRef = asset.blobRef();
+    if (oldBlobRef != null) {
+      if (writePolicySelector.select(asset, writePolicy) == WritePolicy.ALLOW_ONCE) {
```

Comment:
These checks feel linty. What about:

    writePolicySelector.select(asset, writePolicy).checkUpdateAllowed();

Notice that the actual policy definitions already exist as javadocs in the WritePolicy enum, this would consolidate it.